koku/koku/trino_database.py

```
@@ -78,6 +84,130 @@ def query_id(self) -> t.Optional[str]:
+
+    def _retry(callable: t.Callable):
+        @functools.wraps(callable)
+        def wrapper(*args, **kwargs):
+            context = kwargs.get("context", extract_context_from_sql_params(kwargs.get("sql_params", {}))) or None
+            for attempt in range(retries + 1):
+                try:
+                    LOG.debug(f"Attempt {attempt + 1} for {callable.__name__}")
+                    return callable(*args, **kwargs)
+
+                except retry_on as ex:
+                    LOG.debug(f"Exception caught: {ex}")
+
+                    # Verify if the exception contains "NoSuchKey"
+                    # If it does, retry the operation
+                    violation = NoSuchKeyViolation(ex)
+                    if violation.is_no_such_key_violation and attempt < retries:
+                        LOG.warning(
+                            log_json(
+                                msg=f"{log_message} (attempt {attempt + 1})",
+                                context=context,
+                                exc_info=ex,
+                            )
+                        )
+                        backoff = min(2**attempt, max_wait)
+                        jitter = random.uniform(0, 1)
+                        delay = backoff + jitter
+                        LOG.debug(f"Sleeping for {delay} seconds before retrying")
+                        time.sleep(delay)
+                        continue
+
+                    # If it is NoSuchKey and all attempts are over, raise TrinoNoSuchKeyException
+                    if violation.is_no_such_key_violation:
+                        LOG.error(
+                            log_json(
+                                msg=f"Failed execution after {attempt + 1} attempts",
+                                context=context,
+                                exc_info=ex,
+                            )
+                        )
+                        raise TrinoNoSuchKeyException(str(violation))
```

This should raise the original exception on the last failed attempt instead of a `TrinoNoSuchKeyException`.
```suggestion
```